### ISSUE TITLE
void methods and function in null safety

### ISSUE THREAD

**Issue by Renesanse** [dart-lang/sdk#46250]
Right now, void methods returning nil instead of nothing. Analyzer and compiler doesn't handle this behaviour. Here is example: 
```dart
void a(int c) {
  print(c);
}

final c = Completer<void>();

void main() async {
  
  final aValue = (a(1) as dynamic).hashCode;
  // a = nil
  final d = c.future.then((value) {
    //nil
    (value as dynamic).hashCode;
    return value;
  });
  //dynamic d
  // d = nil
  print(d.hashCode);
}
```
dart : 2.14.0-136.0.dev

**Comment by eernstg** [comment]
A method with return type `void` can return anything whatsoever, but a plain `return;` and reaching the end of the function causes `null` to be returned. This is specified and working. What is it that doesn't get handled?

**Comment by lrhn** [comment]
I think this is working as intended.

All Dart functions return *some* value. If you don't write a `return e;`, that value is going to be `null`.
You *shouldn't* call methods on expressions of type `void`, but you *can* (still shouldn't) cast the value of that expression to something else.
The `(a(1) as dynamic)` calls `a`, which returns `null` at type `void`, and then casts that value to `dynamic`. That's all valid.
The `(value as dynamic)` is the same, casting a `void`-typed expression to `dynamic`.
The type of `d` is likely going to be `Future<void>` (because inference infers the value type from the `return value`, and `value` has type `void`).

So, in short, `void` methods do return something, usually `null` (but can return anything, it's the receiver's responsibility to not use the value). You can cast a value with static type `void` to anything using `as` (statically, it might throw at runtime if it isn't actually that type).

**Comment by devoncarew** [comment]
Thanks for the report! Closing as working as intended.
